### DISCUSSION TITLE
Code Generation Alternate Implementation

### DISCUSSION THREAD

**Discussion by asglover** [teddykoker/e3nn.c#9] [Show and tell]
Hi Teddy! 

This is a super cool project, @mitkotak  pointed me towards this repo a week or two ago as we were discussing the idea of code generation for specific kernels. You have a really clean implementation. I am doing the same thing for my project, with the eventual goal of doing code generation for a CUDA kernels. As a test and building block, I wanted to see if I could generate a tp.c file that would work in your framework. I was able to get a tp_v2.c file that passes all the tests, and offers a mild speedup, at least on Perlmutter's CPU (AMD EPYC 7713 64-Core Processor) .  I did a few things differently and wanted to share them incase you find them interesting. 

A plot of the relative speedup I was able to get is below. 

![image](https://github.com/user-attachments/assets/1c35963c-72a6-45e0-bf9c-4b36187ac994)

Let's use `(1,1)->1` as the example to compare the code gen: 

tp.c 
```
void tp_1_1_1(const float* input1, const float* input2, float* output) {
    output[0] = 0.7071067811865474 * input1[1] * input2[2] + -0.7071067811865474 * input1[2] * input2[1];
    output[1] = -0.7071067811865474 * input1[0] * input2[2] + 0.7071067811865474 * input1[2] * input2[0];
    output[2] = 0.7071067811865474 * input1[0] * input2[1] + -0.7071067811865474 * input1[1] * input2[0];
}
```

tp_v2.c
```
// This is an automatically generated kernel for performing a tensor product and contraction
void tp_v2_1_1_1 (
    const float* in_1, 
    const float* in_2, 
    float* out)
{   
    // Define CG Coefficients 
    const float C_0 = 0.70710677;

    // Define Input Values
    float in_1_0 = in_1[0];
    float in_1_1 = in_1[1];
    float in_1_2 = in_1[2];
    float in_2_0 = in_2[0];
    float in_2_1 = in_2[1];
    float in_2_2 = in_2[2];

    // Define Output Accumulators 
    float out_0 = 0;
    float out_1 = 0;
    float out_2 = 0;

    // Perform Outer Products  
    const float op_0_1 = in_1_0 * in_2_1;
    const float op_0_2 = in_1_0 * in_2_2;
    const float op_1_0 = in_1_1 * in_2_0;
    const float op_1_2 = in_1_1 * in_2_2;
    const float op_2_0 = in_1_2 * in_2_0;
    const float op_2_1 = in_1_2 * in_2_1;

    // Perform Contractions
    out_2 += op_0_1 * C_0;
    out_1 -= op_0_2 * C_0;
    out_2 -= op_1_0 * C_0;
    out_0 += op_1_2 * C_0;
    out_1 += op_2_0 * C_0;
    out_0 -= op_2_1 * C_0;

    // Write Output
    out[0] = out_0; 
    out[1] = out_1; 
    out[2] = out_2; 
}
```

The v2 version seems pretty wordy and unnecessary, but I working around the fact that in the function header provided, the input and output pointers aren't declared as `restrict` so technically they could be aliased, so standards compliant assembly code should have to read in `input1[0]` from memory each time it is referenced. I think declaring the pointers as `restrict` should improve this condition and potentially make the explicit reads and writes in the prologue and epilogue unnecessary. I tried to use compiler flags `-fgcse-las` on `tp.c` to force it to perform this optimization and close the performance gap, but I could only improve the gap, not close it. The lack of ‘restrict’ might also be eliminating the opportunity for some instruction level parallelism which is explicitly enabled in the more verbose code. 

Another minor thing I did was "coalesce" constants that are the same up to numerical precision and also only keep the positive version. I know for the numerically different coefficients it should help. I think for sign it should help, although it depends on the compiler a bit. I'd have to look at the assembly to tell. 

This approach was initially slower at really large decompositions because it inherently puts more register pressure unless re-ordered so using the flag `-fsched-pressure` is necessary to improve this by rescheduling the operations in a way that I think at least ideologically should be like diagonalizing a matrix before doing a sparse matrix multiply. 

 I also added `-ffp-contract=fast` to try and bias the compiler to produce FMAs. 

I tried to push my code to a branch to share, but I don't have access currently. I'd be happy to share my source and discuss if you're interested. Let me know if you're able to use these ideas to speed up your implementation.

**Comment by mitkotak** [comment]
Doesn't it let you open a PR after a fork ?

**Reply by asglover** [reply]
I’ll give it another shot tonight, I’m just heading out of the office now. It’s probably my mistake

**Comment by teddykoker** [comment]
Hi @asglover, I'm glad you are interested in the work and I'm eager to try your approach! As @mitkotak mentioned you should be able to create a branch on a [fork](https://github.com/teddykoker/e3nn.c/fork) of the repository and then create a PR from there, but let me know if you are still having issues.

I think making the input/output pointers `restrict` could make sense if it leads to efficiency improvements. If I'm understanding correctly this would only be an issue if you try to do a tensor product of tensor against itself (i.e. tensor square) - but this is probably a reasonable trade off if the resulting product is faster. We can definitely iterate on these ideas once the PR is up :)

**Comment by mitkotak** [comment]
Can you try running these results on your Mac as well and see what we get ?

**Reply by mitkotak** [reply]
Also would be nice to have a separate plot for the different optimizations you tried and how they are getting stacked up.

**Comment by asglover** [comment]
Okay, I forked it and created a pr. In my pr, I added a lot of files so that I could have a `tensor_product_v4` and directly compare it to v3 so that it wasn't causing any problems, but I think you probably would rather just reflect any positive changes into your existing implementation. I shared the tp_v2.c file but not the actual code gen, but my code gen pulls from a slightly different source (e3nn.o3.wigner_3j) for its coefficients and I'm using jinja2 as the template engine. I saw jinja in your requirements .txt but wasn't sure it was working. I can share the details of the actual codegen as well, but it is nothing special.

**Reply by asglover** [reply]
I was thinking about the CG coefficient stability, thing, there is a lot of information here about how to perform these cacluations in a way that numerical stability conscious way. 

https://spherical.readthedocs.io/en/main/

also, I think that somehow the calculations performed by e3nn and e3nn jax are more stable somehow? Maybe as a shortcut, if you know that certain values should have identical values, based on the algorithm, you could write the same value to all the coefficient locations. 

That would help in situations like: 

```
void tp_0_1_1(const float* input1, const float* input2, float* output) {
    output[0] = 0.9999999999999999 * input1[0] * input2[0];
    output[1] = 1.0 * input1[0] * input2[1];
    output[2] = 0.9999999999999999 * input1[0] * input2[2];
}
``` 

Which I think could be a drag on performance. It's requiring at least 1 extra FMA's on top of what should be be 3 FMAs. Even adding a check to force numbers-close-to-1 -> 1 would probably be a good check. Maybe they get  rounded to 1 when they are converted to floats by the compiler with the -sse flag, but you could also put an `f` behind them to guarantee they get interpreted as floats for architectures which don't support that flag.

**Comment by asglover** [comment]
RE: Benchmarking, I think there is a bit of noise, and I'm also trying to measure a small difference, but the shape of the relative performance  curve varies a bit run to run, so probably some kind of multi runs and averaging setup would be necessary to be able to really measure and performance improvements.

I'm not going to try to do rigorous benchmarking the different compiler flags. I tried a few different combinations of adding and removing things, but the overall trend is still there on perlmutter, ~80% so I don't think the flags are causing the relative difference. 

I really need to put this into a profiler to get a better understand of what drives the performance. 

I'll run it on my Mac also just to make sure I'm not getting some kind of weird platform dependent effect. 

I compiled with CLANG instead of GCC so had to use a different set of flags, listed below, and got a different trend, but still the idea that something about the approach yields relative performance. 

![image](https://github.com/user-attachments/assets/60cc1c64-fa2c-4d9d-94dd-cc3a05996352)

compiler flags: 
```
CC=gcc
CFLAGS=-O3 -march=native -funroll-loops 
```

I know it says gcc, but it aliases to clang 
```
(e3nn) austinglover@Austins-MacBook-Pro-3346 e3nn.c % gcc --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by teddykoker** [comment]
@asglover @mitkotak lets move this discussion to #10
